src/DotNetty.Codecs.Http/WebSockets/Extensions/WebSocketServerExtensionHandler.cs

```diff
@@ -88,31 +89,33 @@ public override Task WriteAsync(IChannelHandlerContext ctx, object msg)
                         extensionData.Name, extensionData.Parameters);
                 }
 
-                continuationAction = promise =>
-                {
-                    if (promise.Status == TaskStatus.RanToCompletion)
-                    {
-                        foreach (IWebSocketServerExtension extension in this.validExtensions)
-                        {
-                            WebSocketExtensionDecoder decoder = extension.NewExtensionDecoder();
-                            WebSocketExtensionEncoder encoder = extension.NewExtensionEncoder();
-                            ctx.Channel.Pipeline.AddAfter(ctx.Name, decoder.GetType().Name, decoder);
-                            ctx.Channel.Pipeline.AddAfter(ctx.Name, encoder.GetType().Name, encoder);
-                        }
-                    }
-                    ctx.Channel.Pipeline.Remove(ctx.Name);
-                };
-
                 if (headerValue != null)
                 {
-                    response.Headers.Set(HttpHeaderNames.SecWebsocketExtensions, headerValue);
+                    responseHeaders.Set(HttpHeaderNames.SecWebsocketExtensions, headerValue);
                 }
+
+                Task task = base.WriteAsync(ctx, msg).AsTask();
+                task.ContinueWith(this.upgradeCompletedContinuation, ctx, TaskContinuationOptions.ExecuteSynchronously);
```

Comment:
@StormHub, check this out. Aligned with netty
https://github.com/netty/netty/blob/00afb19d7a37de21b35ce4f6cb3fa7f74809f2ab/codec-http/src/main/java/io/netty/handler/codec/http/websocketx/extensions/WebSocketServerExtensionHandler.java#L118